addons/content-index/src/main/java/org/commonjava/indy/content/index/IndexingContentManagerDecorator.java

```diff
@@ -371,7 +371,13 @@ public Transfer getIndexedTransfer( final StoreKey storeKey, final StoreKey topK
 
         if ( storePath != null )
         {
-            Transfer transfer = delegate.getTransfer( storePath.getStoreKey(), path, op );
+            StoreKey actualStorageKey = storePath.getOriginStoreKey();
```

Comment:
Still confused. The storePath is the value object in the cache. And the only two places where we "put" to the cache are in indexPathInStores. Both are put a value without an originKey. So the storePath.getOriginStoreKey() will always return null. Am I missing something?

Reply:
@ruhan1 you can check ContentIndexManager.indexPathInStores(). When a group case here, the originKey will be the concrete one.

Reply:
When it's stored in the cache, the main key is the one we give it. We index at the concrete repo level and at the group level, which means the main key is sometimes the group key for some of these entries.

However, the origin key is always taken from the Transfer object, which means it is always the key of the concrete store where the file is located.

When we retrieve the indexed entry for a path in a group, we look it up by the group key. This corresponds to the main key used in the IndexedStorePath instance, which allows us to look it up. However, those entries also should contain the origin key information. If the content was generated / aggregated, then the origin key might also be the group key...but otherwise, it should point to the concrete store.

If we always use the main key (even if that comes from the IndexedStorePath instance), then we will always look in the group's storage for the content...the place where we only store aggregated content. This means the group-level indexing will only work for aggregated content...since in all other cases we're constructing a Transfer based on the retrieved index entry that points to the group storage, when we need to look at the concrete repo storage where the file actually lives.

We've been covering this with the indexing in the group membership, but it's slower. That's because each member is traversed in order...and if the indexed entry doesn't exist for that repo, the repo's storage must be checked instead. If we don't check both before going to the next member repo, we risk serving content that should have been "hidden" by content in an earlier member, which simply wasn't indexed yet.

The more I look at this, the more certain I am that this is the case, and it's a major cause of performance problems.

Reply:
The real key here is knowing that topKey doesn't affect originKey. Origin key is always taken from `LocationUtils.getKey( transfer )` inside ContentIndexManager's code.

Reply:
Maybe we can make it clear by changing value type to StoreKey? e.g.,
@Inject
private CacheHandle<IndexedStorePath, StoreKey> contentIndex;
The path is in IndexedStorePath so we don't need it in the value. It is to say: if you want <store x, path p>, go find it in store y.